README.md
# EnnouncementGen

A Google Apps Script solution for building a weekly email announcement system.

## Features

* Creates a simple HTML email of submitted announcements.
* Checks that announcements have been approved in the sheet and not already sent.
* Support for recurring announcements (will send on endDate).
* Uses showdown.js to convert Markdown to HTML.

## Known Issues

* Cancelling a "Send command" will still cause announcements to update their "Sent?" status. Make sure you use the undo button if this happens.


## Instructions

This script is intended to be used with a Google Sheet.

### Getting Started
1. Upload the scripts to a Google Sheets file.
2. Create a settings.js file from settings.js.dist.
3. In settings.js you can customize the categories, recipients, who to CC the message to, and the weekly request notification.
4. Reload the sheet to see the menu.
5. Click Ennouncements -> Setup Sheet to create and link the form as well as format the sheet correctly.

### Sending Ennouncements

#### Automatic Method (easiest)

1. Click Install Triggers to schedule the reminder and weekly ennouncement email automatically and lock the form at the deadline each week.
2. You will only have to approve submissions before the email is automatically sent out.
3. If you miss one, you will get a reminder to approve and send the email manually.

#### Manual Method

1. As submissions come in, you will have to approve them by typing a  "Y" or "N" in the `Approved?` column.
2. To preview the email, click "Test Only" and you will see the final output in a popup box.
3. To send the email click "Send Ennouncements" and you will be prompted to confirm who you'd like to send it to.
   1. **NOTE:** If you click no, make sure you undo the updated `Sent?` statuses. This is a current issue.
4. To send a reminder email asking users for their submissions, click "Send Request." The "Test Request" button will let you preview this email. It automatically includes a link to the linked submission form.

CHANGELOG.md
# CHANGELOG

## 4.3

* Added a sorting option to post the newest announcements first.

## 4.2

* Added additional options to make approval and auto-locking the form optional.

## 4.1

* Moved changelog to separate file.
* Separated settings.js and removed old dependency on urlshortener.
* Added additional instructions to README.

## 4.0

* Added showdown to convert text to Markdown https://github.com/showdownjs/showdown
* Licensed under MIT.

## 3.4

* Updated for 2019-2020.
* Made TriggerManager to be able to automatically install time-based triggers to manage the sending of requests and the announcements as well as locking/unlocking the form.

## 3.3

* Added the ability to lock the form. (Could be configured to a trigger.)
* sendEnnouncementPrompt (auto sending) will now email the owner/user of the sheet if there are pending announcements.

## 3.2

* Added methods to send without prompting.

## 3.1

* Refactored to make a singular loop instead of one to build the array and one to build the HTML.

## 3.02

* Made a fix to Form linking and switches to sheet

## 3.01

* Added new form features to menu.

## 3.0

* Updated for 2018-2019.
* Can now properly create and link the Google Form from scratch and format the Sheet.
* Column references are now based on the proper order.

## 2.0

* Now supports categories.

## 1.3

* FIXED: Month in subject and message needed +1 because .getMonth() starts at 0.

## 1.2

* FIXED: Selecting "No" in send prompt.
* Added class email lists and staff to recipients list.

## 1.1

* FIXED: Contact email HTML.
* Added more comments and renamed some variables for clarity.
* Added a line to the beginning of the email with the date in ordinal format (th, st, etc).

## 1.01

* FIXED: multiline support with white-space: pre-wrap in message body.
* FIXED: Reads proper values in new sheet.

## 1.0

* Initial Release


tests.js
function testEnnouncement() {
    //Only does a test run of which ennouncements to send and shows the HTMLOutput.
    var ui = SpreadsheetApp.getUi();
    var ss = SpreadsheetApp.getActiveSpreadsheet();
    var htmlBody = buildEnnouncement(ss, true);
    ui.showModalDialog(HtmlService.createHtmlOutput(htmlBody), 'Ennouncement Output');
}

function testRequest() {
    //Only does a test run of the request message to send and shows the HTMLOutput.
    var ui = SpreadsheetApp.getUi();
    var htmlBody = buildRequestHtml();
    ui.showModalDialog(HtmlService.createHtmlOutput(htmlBody), 'Ennouncement Output');
}

function testARA() {
    Logger.log(areSubmissionsApproved());
}

FormManager.js
function buildForm() {
  var scriptProperties = PropertiesService.getScriptProperties();
  //Creates the form with all of the correct settings.
  //var userProperties = PropertiesService.getUserProperties();

  var form = FormApp.create("Ennouncements");
  form.setTitle("Ennouncements")
    .setAcceptingResponses(true)
    .setCollectEmail(true)
    .setShowLinkToRespondAgain(true)
    .setLimitOneResponsePerUser(false)
    .setDescription("Please submit your E-announcement for approval here.\n\
Messages that are approved will be printed as-is, so please proofread them before sending.\n\
Messages should be sent by Fridays at 3:15 PM.");

  //Add sample image.
  var img = DriveApp.getFileById('172r0YeajO3HRzy5Do-KK_lqPybFO7KPC');
  form.addImageItem()
    .setTitle("Sample Ennouncement")
    .setImage(img.getBlob())
    .setAlignment(FormApp.Alignment.CENTER);

  //Categories
  var category = form.addListItem();
  category.setTitle("Category")
    .setHelpText("Priority and Office is for use of office staff only.")
    .setChoices([
      category.createChoice("Priority"),
      category.createChoice("Office"),
      category.createChoice("Events"),
      category.createChoice("Clubs"),
      category.createChoice("Other")])
    .setRequired(true);

  //Title
  var textValidation = FormApp.createTextValidation()
        .requireTextLengthLessThanOrEqualTo(50)
        .setHelpText("Title must 50 characters or less.")
        .build();
  
  form.addTextItem()
    .setTitle("Announcement Title")
    .setHelpText("What would you like the title of your announcement to be? (50 characters or less)")
    .setValidation(textValidation)
    .setRequired(true);

  //Body Text
  //Formatting Tutorial
  form.addSectionHeaderItem()
    .setTitle("Announcement Text")
    .setHelpText("You can now format your text for emphasis by wrapping parts of your message in certain characters.\n\
*text* = italic text\n\
**text** = bold text\n\
__text__ = underline text\n\
\n\
Learn more here: https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet");
  
  //Announcement Text"
  var textValidation =  FormApp.createParagraphTextValidation()
        .requireTextLengthLessThanOrEqualTo(500)
        .setHelpText("Messages must be 500 characters or less.")
        .build();
  form.addParagraphTextItem().setTitle("Announcement Text")
    .setHelpText("The content of your announcement. (500 characters or less) Need to shorten a link? Use https://goo.gl/ or https://bit.ly/")
    .setValidation(textValidation)
    .setRequired(true);

  //End Date
  form.addDateItem()
    .setHelpText("(OPTIONAL) If this is a multi-week announcement, please enter the LAST DAY it is applicable to. Leaving this blank = one-time announcement. Except in special cases, please be courteous and do not keep an ennouncement running for more than a month.")
    .setTitle("End Date");

  //Contact Email
  
  var textValidation = FormApp
        .createTextValidation()
        .requireTextIsEmail()
        .setHelpText("Not a valid E - mail address.")
        .build();
  form.addTextItem()
    .setHelpText("(OPTIONAL) If you want to include a contact email after your message.")
    .setTitle("Contact Email")
    .setValidation(textValidation);

  form.setAcceptingResponses(true)

  scriptProperties.setProperty('ennouncementFormId', form.getId());
  Logger.log(form.getPublishedUrl());
  return form;
}

function getSavedFormId() {
  //Returns the saved formId in the ScriptProperties
  var scriptProperties = PropertiesService.getScriptProperties();
  return scriptProperties.getProperty('ennouncementFormId');
}

function linkFormToSheet() {
  //Check if a form is already saved and links it. Else it creates a new one.
  //Returns the final sheet.
  var scriptProperties = PropertiesService.getScriptProperties();
  var ss = SpreadsheetApp.getActiveSpreadsheet();

  if (!ss.getFormUrl() && scriptProperties.getProperty('ennouncementFormId')) {
    linkSavedForm()
  }
  else {
    linkNewForm()
  }
  var responseSheet = renameResponseSheet();
  ss.setActiveSheet(responseSheet);
  return responseSheet;
}

function linkNewForm() {
  var scriptProperties = PropertiesService.getScriptProperties();
  var ss = SpreadsheetApp.getActiveSpreadsheet();

  var form = buildForm();
  scriptProperties.setProperty('ennouncementFormId', form.getId());

  return form.setDestination(FormApp.DestinationType.SPREADSHEET, ss.getId());
}

function linkSavedForm() {
  var ss = SpreadsheetApp.getActiveSpreadsheet();
  var form = FormApp.openById(getSavedFormId());
  return form.setDestination(FormApp.DestinationType.SPREADSHEET, ss.getId());
}

function renameResponseSheet() {
  var ss = SpreadsheetApp.getActiveSpreadsheet();
  var sheets = ss.getSheets();
  for (i = 0; i < sheets.length; i++) {
    var name = sheets[i].getName();
    if (name.indexOf("Form Responses") == 0) {
      break;
    }
  }
  return sheets[i].setName("Ennouncements");
}

function lockForm() {
  if (isAutoLockFormEnabled()) {
    var form = FormApp.openById(getSavedFormId());
    form.setCustomClosedFormMessage("The ennouncement form has been closed for this week. Please wait until next week to submit your ennouncement. \n Ennouncements are due by 3:15pm the week before.");
    form.setAcceptingResponses(false);
  }
}

function unlockForm() {
  if (isAutoLockFormEnabled()) {
    var form = FormApp.openById(getSavedFormId());
    form.setAcceptingResponses(true);
  }
}

function unlinkFormAndDeleteAll() {
  var form = FormApp.openById(getSavedFormId());
  var ss = SpreadsheetApp.getActiveSpreadsheet();
  var sheet = ss.getSheetByName("Ennouncements");
  form.removeDestination();
  form.deleteAllResponses();
  var TODAY = new Date();
  sheet.setName("Archive " + TODAY.toString());
}

SheetManager.js
/*
Name: SheetManager.gs
Author: Alex Wolfe
Description: Provides functions for managing the Google Sheet.
*/

function sheetSetup() {
  //Configures the desired settings for the sheet.
  var ss = SpreadsheetApp.getActiveSpreadsheet();
  var sheet = linkFormToSheet();
  saveSheetId();

  //Bold and freeze top row.
  var firstRow = sheet.getRange(1, 1, 1, sheet.getMaxColumns());
  firstRow.setFontWeight("bold");
  sheet.setFrozenRows(1);


  //Delete extra columns
  var maxColumns = sheet.getMaxColumns();
  var lastColumn = sheet.getLastColumn();
  if (lastColumn == 7) {
    sheet.getRange(1, 8).setValue("Approved");
    sheet.getRange(1, 9).setValue("Sent?");
    lastColumn += 2;
  }
  if (maxColumns > lastColumn) {
    sheet.deleteColumns(lastColumn + 1, maxColumns - (lastColumn));
  }


  //Create formatting rules for Y/N/R.
  var range = sheet.getRange(1, 8, sheet.getMaxRows(), 9);
  var rules = sheet.getConditionalFormatRules();

  rules.push(
    SpreadsheetApp.newConditionalFormatRule()
      .whenTextEqualTo("Y")
      .setBackground("#00AA00")
      .setRanges([range])
      .build());
  rules.push(
    SpreadsheetApp.newConditionalFormatRule()
      .whenTextEqualTo("N")
      .setBackground("#AA0000")
      .setRanges([range])
      .build());
  rules.push(
    SpreadsheetApp.newConditionalFormatRule()
      .whenTextEqualTo("R")
      .setBackground("#FFCC00")
      .setRanges([range])
      .build());
  
  //Highlight recurring dates that are past.
  rules.push(
    SpreadsheetApp.newConditionalFormatRule()
      .whenDateBefore(SpreadsheetApp.RelativeDate.TODAY)
      .setBackground("#AA0000")
      .setRanges([sheet.getRange(1, 6, sheet.getMaxRows(), 6)])
      .build());

  sheet.setConditionalFormatRules(rules);
}

function copyRowToSheet(sourceSheet, sourceRow, destSheet, destRow) {
  sourceSheet = SpreadsheetApp.getActiveSheet();
  destSheet = SpreadsheetApp.getActiveSheet();
  var sourceRow = sourceSheet.getRange(sourceRow, sourceRow, 1, sourceSheet.getMaxColumns()).
    destSheet.appendRow(rowContents)
}

function saveSheetId() {
  var scriptProperties = PropertiesService.getScriptProperties();
  scriptProperties.setProperty('ennouncementSheetId', SpreadsheetApp.getActiveSpreadsheet().getId());
}

function getSavedSheetId() {
  var scriptProperties = PropertiesService.getScriptProperties();
  return scriptProperties.getProperty('ennouncementSheetId');
}

TriggerManager.js
function installTriggers() {
  var userProperties = PropertiesService.getUserProperties(); 
  var properties = userProperties.getProperties();
  
  //Check if each of the triggers have an ID (are currently installed)
  var sendTriggerID = properties["sendTrigger"];
  var requestTriggerID = properties["requestTrigger"];
  var lockTriggerID = properties["lockTrigger"];
  var unlockTriggerID = properties["unlockTrigger"];
  
  if (!sendTriggerID && !requestTriggerID && !lockTriggerID && !unlockTriggerID) {
    // Send every week on Monday at 0800.
    var sendTrigger = ScriptApp.newTrigger('sendEnnouncementAuto')
    .timeBased()
    .onWeekDay(ScriptApp.WeekDay.MONDAY)
    .atHour(8)
    .create();
    
    userProperties.setProperty("sendTrigger", sendTrigger.getUniqueId()); //Remember the ID for the user to delete it later.
    
    // Request submissions every Thursday at 1200.
    var requestTrigger = ScriptApp.newTrigger('sendRequestAuto')
    .timeBased()
    .onWeekDay(ScriptApp.WeekDay.THURSDAY)
    .atHour(12)
    .create();
    
    userProperties.setProperty("requestTrigger", requestTrigger.getUniqueId());
    
    // Lock the form every Friday at 1515.
    var lockTrigger = ScriptApp.newTrigger('lockForm')
    .timeBased()
    .onWeekDay(ScriptApp.WeekDay.FRIDAY)
    .atHour(15)
    .nearMinute(15)
    .create();
    
    userProperties.setProperty("lockTrigger", lockTrigger.getUniqueId());
    
    // Unlock the form every Monday at 0800.
    var unlockTrigger = ScriptApp.newTrigger('unlockForm')
    .timeBased()
    .onWeekDay(ScriptApp.WeekDay.MONDAY)
    .atHour(8)
    .create();
    
    userProperties.setProperty("unlockTrigger", unlockTrigger.getUniqueId());
  }
  else {throw ("Triggers are already installed. Stopping.");}
}

function deleteTriggers() {
  var userProperties = PropertiesService.getUserProperties();
  var properties = userProperties.getProperties();
  var sendTriggerID = properties["sendTrigger"];
  var requestTriggerID = properties["requestTrigger"];
  var lockTriggerID = properties["lockTrigger"];
  var unlockTriggerID = properties["unlockTrigger"];
  
  if (sendTriggerID && requestTriggerID && lockTriggerID && unlockTriggerID) {
    deleteTrigger(sendTriggerID);
    deleteTrigger(requestTriggerID);
    deleteTrigger(lockTriggerID);
    deleteTrigger(unlockTriggerID);
    
    userProperties.deleteProperty("sendTrigger");
    userProperties.deleteProperty("requestTrigger");
    userProperties.deleteProperty("lockTrigger");
    userProperties.deleteProperty("unlockTrigger");
  }
  else {throw ("Triggers are not currently installed. Stopping.");}
}

function deleteTrigger(triggerId) {
  // Loop over all triggers.
  var allTriggers = ScriptApp.getProjectTriggers();
  for (var i = 0; i < allTriggers.length; i++) {
    // If the current trigger is the correct one, delete it.
    if (allTriggers[i].getUniqueId() === triggerId) {
      ScriptApp.deleteTrigger(allTriggers[i]);
      break;
    }
  }
}

EnnouncementGen.js
/*
  Name:     Ennouncement Generator
  Author:   Alex Wolfe
  Version:  4.2

  License:  GPLv3 (https://www.gnu.org/licenses/gpl.html)

  Features:
  -Creates a simple HTML email of submitted announcements.
  -Checks that announcements have been approved in the sheet and not already sent.
  -Support for recurring announcements (will send on endDate).
  -Uses showdown.js to convert Markdown to HTML.

  ISSUES:
  -Cancelling a "Send command" will still cause announcements to update their "Sent?" status. Make sure you use the undo button if this happens.

*/
var showdownConverter;

function getVersion() {
  return "4.1"
}


function onOpen() {
  //Add a menu to the spreadsheet.
  var ui = SpreadsheetApp.getUi();
  ui.createMenu('Ennouncements')
    //.addItem('Test and Send Ennouncement', 'testAndSendEnnouncement') //Test and Send appears to be broken at the moment.
    .addItem('Send Only', 'sendEnnouncement')
    .addItem('Test Only', 'testEnnouncement')
    .addItem('Send to Self', 'sendToSelf')
    .addItem('Send REQUEST', 'sendRequest')
    .addItem('Test REQUEST', 'testRequest')
    .addSeparator()
    .addItem('Setup Sheet', 'sheetSetup')
    .addItem('Unlink Form and Archive Responses', 'unlinkFormAndDeleteAll')
    .addSeparator()
    .addItem('Install Triggers', 'installTriggers')
    .addItem('Delete Triggers', 'deleteTriggers')
    .addToUi();
}

function prepareRow(row) {
  /*
    [i][0] Datestamp (unused)
    [i][1] Author's email (unused)
    [i][2] Category
    [i][3] Title
    [i][4] Body Text
    [i][5] End Date (optional)
    [i][6] Contact Email (optional)
    [i][7] Approved? (Y/N)
    [i][8] Sent? (Y/N/R)
  */
  //If category doesn't match, set to Other.
  var TODAY = new Date();
  TODAY.setHours(0);
  TODAY.setMinutes(0);
  TODAY.setSeconds(0);
  TODAY.setMilliseconds(0);

  if (getCategories().indexOf(row[2]) == -1) { row[2] = "Other" };

  row[3].replace(/(<([^>]+)>)/ig, ""); //clear HTML tags
  row[4].replace(/(<([^>]+)>)/ig, ""); // clear HTML tags

  var showdown = getConverter();
  var html = showdown.makeHtml(row[4]);
  row[4] = html;

  //EndDate Handling
  if (row[5] == "") {
    row[5] = new Date(TODAY); //If no tempEndDate, set it to be TODAY.
  } else {
    row[8] = "R"
    row[5] = new Date(row[5].getTime());
  }
  row[5].setHours(0);
  row[5].setMinutes(0);
  row[5].setSeconds(0);
  row[5].setMilliseconds(0);

  return row;
}

function processRow(row, i, testOnly) {
  //Processes the given row, returning whether or not the row will be sent (for counting purposes).
  var TODAY = new Date();
  TODAY.setHours(0);
  TODAY.setMinutes(0);
  TODAY.setSeconds(0);
  TODAY.setMilliseconds(0);

  var categories = getCategories();

  //Process the row.
  var tempRow = prepareRow(row);
  var tempEndDate = tempRow[5];
  var isApproved = !isApprovalRequired();
  if (tempRow[7] == "Y") { isApproved = true; }

  //Logger.log(tempRow);

  var isSent = false;
  var isRecurring = false;
  var toSend = false;

  if (tempRow[8] == "Y") { isSent = true; }
  else if (tempRow[8] == "R") { isRecurring = true; }
  //Done processing row.

  //Build the HTML for the row
  if (isSent || !isApproved || TODAY.getTime() > tempEndDate.getTime()) {
    //If already sent, or not approved, or past the tempEndDate then don't add to HTML.
  }
  else {
    //announcementCount++;
    var tempCategory = categories.indexOf(tempRow[2]);
    categoryHtml[tempCategory] += buildRowHtml(tempRow);
    toSend = true;
  }

  //TODO: Write sent value AFTER confirming send?
  if (!testOnly) {
    //Logger.log("Not in testing mode. Writing sent value.");
    var statusCell = announcementSheet.getRange(2 + i, 9, 1);
    if ((isApproved && isRecurring && (TODAY.getTime() >= tempEndDate.getTime())) || (!isRecurring && isApproved)) {
      //End of recurring announcement, or sending a one-time event.
      statusCell.setValue("Y");
    }
    else if (isApproved && isRecurring) {
      //Unfinished recurring announcement
      statusCell.setValue("R");
    }
    else {
      //Do not send.
      statusCell.setValue("N");
    }
  }
}

function buildRowHtml(row) {
  //Builds the individual HTML for a row and returns it.
  var tempHtml = '';
  var tempCategory = getCategories().indexOf(row[2]);
  tempHtml += '<u><h3>' + row[3] + '</h3></u>\n';
  tempHtml += "<p style='white-space:pre-wrap;'>" + row[4] + "</p>\n";
  if (row[6] != "") { tempHtml += '<em>For more details, please contact: <a href="mailto:' + row[6] + '">' + row[6] + '</a></em>\n'; }
  tempHtml += "<hr />"
  return tempHtml;
}

function buildEnnouncement(spreadsheet, testOnly) {
  if (arguments.length == 1) {
    var testOnly = false;
  } else {
    var testOnly = testOnly;
  }
  //Make a variable for today and set the time to 00:00h for ease of checking.
  var TODAY = new Date();
  TODAY.setHours(0);
  TODAY.setMinutes(0);
  TODAY.setSeconds(0);
  TODAY.setMilliseconds(0);

  var announcementSheet = spreadsheet.getSheetByName("Ennouncements");
  var maxRows = announcementSheet.getLastRow();

  var announcementArray = announcementSheet.getRange(2, 1, maxRows - 1, 9).getValues();
  var announcementCount = 0;

  //Setup HTML stuff.
  var html = new String();
  var DAYS = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];
  var MONTHS = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'];
  var categories = getCategories();
  //Creates an array to hold the HTML for each category in order.
  var categoryHtml = new Array(categories.length);
  for (i = 0; i < categoryHtml.length; i++) { categoryHtml[i] = ''; }

  html += "<h2>Ennouncements for the Week of " + (TODAY.getMonth() + 1) + "/" + TODAY.getDate() + "/" + TODAY.getYear() + "</h2>\n \
    <p>Thank you students and staff for your submissions, here are your ennouncements for " + DAYS[TODAY.getDay()] + ", " + MONTHS[TODAY.getMonth()] + " " + dateOrdinal(TODAY.getDate()) + ", " + TODAY.getYear() + ".</p>\n \
    </br>\n";

  if (isSortOldToNew()) {
    for (var i = 0; i < announcementArray.length; i++) {
      var toSend = processRow(announcementArray[i], i, testOnly);
      if (toSend)
        announcementCount++;
    }
  }
  else {
    for (var i = announcementArray; i >= 0; i--) {
      var e = processRow(announcementArray[i], i, testOnly);
      if (toSend)
        announcementCount++;
    }
  }
  /*
  Logger.log("Title: "+announcementArray[i][2]);
  Logger.log("Body: "+announcementArray[i][3]);
  Logger.log("isApproved "+isApproved);
  Logger.log("isSent "+isSent);
  Logger.log("tempEndDate "+tempEndDate.getTime());
  Logger.log("isRecurring "+isRecurring);
  */

  if (announcementCount == 0) { throw ("No ennouncements that need sent. Stopping."); }

  for (i = 0; i < categories.length; i++) {
    if (categoryHtml[i] != '') {
      html += '<strong><h1>' + categories[i] + '</h1></strong>'
      html += categoryHtml[i];
    }
  }
  html += getByline();

  //Logger.log(announcementArray);
  return html;
}

function dateOrdinal(d) {
  //Returns a nice ordinal for a given date (d).
  return d + (31 == d || 21 == d || 1 == d ? "st" : 22 == d || 2 == d ? "nd" : 23 == d || 3 == d ? "rd" : "th")
}

function sendToSelf() {
  //Shows the HTMLOutput before sending to your own email address.
  var ss = SpreadsheetApp.getActiveSpreadsheet();
  var RECIPIENTS = Session.getEffectiveUser().getEmail();
  var SUBJECT = getSubject();

  var htmlBody = buildEnnouncement(ss);

  return sendEnnouncementEmail(RECIPIENTS, SUBJECT, htmlBody);
}

function sendEnnouncement() {
  //Sends the ennouncement.
  var ss = SpreadsheetApp.getActiveSpreadsheet();
  var RECIPIENTS = getRecipients();
  var SUBJECT = getSubject();

  var htmlBody = buildEnnouncement(ss);

  return sendEnnouncementEmail(RECIPIENTS, SUBJECT, htmlBody, true);
}

function sendEnnouncementAuto() {
  //Despite the name it actually sends wih NO prompt.
  var ss = SpreadsheetApp.openById(getSavedSheetId());
  var RECIPIENTS = getRecipients();
  var SUBJECT = getSubject();

  if (areSubmissionsApproved()) {
    var htmlBody = buildEnnouncement(ss);
    return sendEnnouncementEmail(RECIPIENTS, SUBJECT, htmlBody, false);
  }
  else {
    var url = ss.getUrl();
    return MailApp.sendEmail(Session.getEffectiveUser().getEmail(), "Ennouncement Problem", "Not all submissions on the sheet were approved, go to: " + url + "\n" +
      "To approve/deny all pending announcements and then manually send using the Ennouncements->Send Only menu.");
  }
}

function sendEnnouncementEmail(recipients, subject, htmlBody, doPrompt) {

  if (doPrompt) {
    var ui = SpreadsheetApp.getUi();
    //Sends the ennouncement email.
    var response = ui.alert("Send Ennouncement", "Do you want to send the ennouncement to " + recipients + "?", ui.ButtonSet.YES_NO);
    if (response == ui.Button.YES) {
      var send = true;
    } else {
      throw ("User chose to not send ennouncement. Stopping.");
    }
  }
  return MailApp.sendEmail({
    to: recipients,
    bcc: getCC(),
    subject: subject,
    htmlBody: htmlBody,
  })
}

function sendRequest() {
  var ui = SpreadsheetApp.getUi();
  var recipients = getRecipients();
  var response = ui.alert("Send REQUEST for Ennouncements", "Do you want to send the REQUEST to " + recipients + "?", ui.ButtonSet.YES_NO);
  if (response == ui.Button.YES) {
    var send = true;
  } else {
    throw ("User chose to not send ennouncement. Stopping.");
  }
  var htmlBody = buildRequestHtml();

  return MailApp.sendEmail({
    to: recipients,
    bcc: getCC(),
    subject: "Request for Ennouncements",
    htmlBody: htmlBody
  });
}

function sendRequestAuto() {
  //Automatically send the request without prompting.
  var htmlBody = buildRequestHtml();
  var recipients = getRecipients();

  return MailApp.sendEmail({
    to: recipients,
    bcc: getCC(),
    subject: "Request for Ennouncements",
    htmlBody: htmlBody
  });
}

function getByline() {
  html = '<br /> \
<p style="color:#C0C0C0; font-size: xx-small"><a href="https://github.com/afwolfe/EnnouncementGenerator">EnnouncementGen</a> by Alex Wolfe v' + getVersion() + '<br /> \
Uses <a href="https://github.com/showdownjs/showdown">showdown.js</a> licensed under <a href="https://www.opensource.org/licenses/MIT">MIT License.</a></p>';
  return html;
}


function areSubmissionsApproved() {
  //Checks if all submissions are approved.
  if (isApprovalRequired()) {
    var ss = SpreadsheetApp.getActiveSpreadsheet();
    var ennouncements = ss.getSheetByName('Ennouncements');
    var maxRows = ennouncements.getLastRow();

    var ennouncementArray = ennouncements.getRange(2, 1, maxRows - 1, 9).getValues();

    for (i = 0; i < ennouncementArray.length; i++) {
      if (ennouncementArray[i][7] == '') {
        Logger.log('false');
        return false;
      }
    }
  }
  return true
}

function getConverter() {
  if (!showdownConverter) {
    showdownConverter = new showdown.Converter();
    showdownConverter.setOption('headerLevelStart', 4);
    showdownConverter.setOption('simplifiedAutoLink', true);
    showdownConverter.setOption('simpleLineBreaks', true);
  }
  return showdownConverter;
}
